tests/integration_tests/functional/test_serial_io.py

```diff
@@ -128,3 +132,65 @@ def test_serial_dos(test_microvm_with_ssh):
                                       "changed from {} to {}." \
                                       .format(before_size,
                                               after_size)
+
+
+def test_serial_block(test_microvm_with_ssh, network_config):
+    """Test that writing to stdout never blocks the vCPU thread."""
+    test_microvm = test_microvm_with_ssh
+    test_microvm.jailer.daemonize = False
+    test_microvm.spawn()
+    # Set up the microVM with 1 vCPU so we make sure the vCPU thread
+    # responsible for the SSH connection will also run the serial.
+    test_microvm.basic_config(
+        vcpu_count=1,
+        mem_size_mib=512,
+        boot_args='console=ttyS0 reboot=k panic=1 pci=off'
+    )
+
+    _tap, _, _ = test_microvm.ssh_network_config(network_config, '1')
+
+    # Configure the metrics.
+    metrics_fifo_path = os.path.join(test_microvm.path, 'metrics_fifo')
+    metrics_fifo = log_tools.Fifo(metrics_fifo_path)
+    response = test_microvm.metrics.put(
+        metrics_path=test_microvm.create_jailed_resource(metrics_fifo.path)
+    )
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+
+    test_microvm.start()
+    ssh_connection = net_tools.SSHConnection(test_microvm.ssh_config)
+
+    # Get an initial reading of missed writes to the serial.
+    fc_metrics = test_microvm.flush_metrics(metrics_fifo)
+    init_count = fc_metrics['uart']['missed_write_count']
+
+    screen_pid = test_microvm.screen_pid
+    # Stop `screen` process which captures stdout so we stop consuming stdout.
+    subprocess.check_call(
+        "kill -s STOP {}".format(screen_pid),
+        shell=True
+    )
+
+    # Generate a random text file.
+    exit_code, _, _ = ssh_connection.execute_command(
+        "base64 /dev/urandom | head -c 100000000 > file.txt"
+    )
+
+    # Function to dump output to terminal
+    def write_to_serial():
+        _, _, _ = ssh_connection.execute_command(
+            "cat file.txt > /dev/ttyS0"
+        )
+
+    # Put it in another thread so we can asynchronously check the metric.
```

Comment:
thread can be replaced with `execute_command` with timeout (or even without timeout since the whole test has a timeout) - if the vcpu gets stuck during `cat bla > /dev/ttyS0` cmd should timeout.

Reply:
We do it on another thread because `execute_command` blocks until the actual command finished running in the guest and returns the output. We want to be able to do other stuff, which you can see a few lines below, while that command is executing.

Reply:
I understand that the extra thread allows parallel execution of another command. But that isn't necessary. The `cat bla > /dev/ttyS0` command either blocks the vcpu or it doesn't, and that can be verified without the extra thread and sleeping.
Simply run the command and expect retcode zero. It will either succeed or time out.

This is not a PR blocker, but I do agree with Alin that we can reduce complexity by dropping the extra thread.

Reply:
Removed the extra thread.